traversal/procedure/ProcedureVertex.java

```diff
@@ -175,124 +189,347 @@ <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> iterateAndFilter
+                    );
+                } else {
+                    return merge(iterate(vertexIters)
+                            .map(pair -> applyPredicatesOnVertices(graphMgr, params, pair.first(), pair.second())
+                                    .mapSorted(
+                                            a -> a.asAttribute().toValue(),
+                                            v -> {
+                                                assert v.isValue() && pair.first().valueType().comparables().contains(v.valueType());
+                                                return attributeVertexTarget(graphMgr, pair.first(), v, order.isAscending());
+                                            },
+                                            order
+                                    )
+                            ), order);
+                }
+            }
+        }
+
+        <ORDER extends Order> Forwardable<KeyValue<ThingVertex, ThingVertex>, ORDER> mergeAndFilterPredicatesOnEdges(
+                GraphManager graphMgr, List<Pair<TypeVertex, Forwardable<KeyValue<ThingVertex, ThingVertex>, ORDER>>> edgeIters,
+                Traversal.Parameters params, ORDER order
+        ) {
+            if (props().predicates().isEmpty()) return merge(iterate(edgeIters).map(Pair::second), order);
+            else {
+                if (iterate(edgeIters).anyMatch(pair -> pair.first().isAttributeType() && pair.first().asType().valueType().equals(STRING))) {
+                    return merge(
+                            iterate(edgeIters).map(pair -> pair.second().filter(a -> checkPredicates(a.key(), params, set()))),
+                            order
+                    );
+                } else {
+                    return merge(iterate(edgeIters)
+                                    .map(pair -> applyPredicatesOnEdges(graphMgr, params, pair.first(), pair.second())
+                                            .mapSorted(
+                                                    a -> new KeyValue<>(a.key().asAttribute().toValue(), a.value()),
+                                                    v -> {
+                                                        AttributeVertex.Value<?> value = v.key().asAttribute().toValue();
+                                                        assert pair.first().valueType().comparables().contains(value.valueType());
+                                                        ThingVertex target = attributeVertexTarget(graphMgr, pair.first(), value, order.isAscending());
+                                                        return new KeyValue<>(target, v.value());
+                                                    },
+                                                    order)
+                                    ),
+                            order);
+                }
+            }
+        }
+
+        private <ORDER extends Order> Forwardable<ThingVertex, ORDER> applyPredicatesOnVertices(
```

Comment:
this is the optimised execution of predicates on a `Attribute` vertex iterator. We can get the biggest Greater-Than predicate and skip forward to that value on the iterator, if the iterator is "Ascending". We can also stop at the smallest Less-Than predicate!

The inverse applies to the "Descending" iterators - we can skip to the smallest Less-Than value, and stop at the largest Greater-Than value. In systems with large numbers of attributes within the type, this can be a huge win!